### PR TITLE
OSSM-1063: Switch to clang

### PR DESCRIPTION
**Pull request by jwendell**

### CODE REVIEW

File: bazel/envoy_internal.bzl [hunk 1]
```diff
@@ -7,7 +7,7 @@ def envoy_copts(repository, test = False):
     posix_options = [
         "-Wall",
         "-Wextra",
-        "-Werror",
+#        "-Werror", FIXME: https://issues.redhat.com/browse/OSSM-1201
```
Comment (oschaaf) [inline bazel/envoy_internal.bzl:10]: We could remove this line?

File: bazel/foreign_cc/luajit.patch [hunk 2]
```diff
@@ -11,15 +11,6 @@ index e65b55e..f0a61dd 100644
  #
  # Use this if you want to force a 32 bit build on a 64 bit multilib OS.
  #CC= $(DEFAULT_CC) -m32
-@@ -36,7 +36,7 @@ CC ?= $(DEFAULT_CC)
- # to slow down the C part by not omitting it. Debugging, tracebacks and
- # unwinding are not affected -- the assembler part has frame unwind
- # information and GCC emits it where needed (x64) or with -g (see CCDEBUG).
--CCOPT= -O2 -fomit-frame-pointer
-+CCOPT= -O2 -fomit-frame-pointer -fPIC -fPIE -fplugin=annobin -fstack-protector-strong -fstack-protector-all
```
Comment (oschaaf) [inline bazel/foreign_cc/luajit.patch:19]: question: this looks like annobin & at least some of the hardening flags work with annobin?

File: bazel/repositories.bzl [hunk 3]
```diff
@@ -821,10 +821,7 @@ def _com_googlesource_chromium_v8():
         name = "com_googlesource_chromium_v8",
         genrule_cmd_file = "@envoy//bazel/external:wee8.genrule_cmd",
         build_file = "@envoy//bazel/external:wee8.BUILD",
-        patches = [
-            "@envoy//bazel/external:wee8.patch",
-            "@envoy//bazel/external:wee8-annobin.patch",
```
Comment (oschaaf) [inline bazel/repositories.bzl:826]: is the plan to re-enable this patch for annobin in a follow up?

File: maistra/bazelrc [hunk 4]
```diff
@@ -0,0 +1,5 @@
+# FIXME: https://issues.redhat.com/browse/OSSM-1201
+build --cxxopt -w
```
Comment (oschaaf) [inline maistra/bazelrc:2]: Do we still need this? (No objections to merging as-is and mopping up later if so, just curious)

File: maistra/run-ci.sh [hunk 5]
```diff
@@ -16,13 +16,12 @@ export BUILD_SCM_REVISION="Maistra PR #${PULL_NUMBER:-undefined}"
 export BUILD_SCM_STATUS="SHA=${PULL_PULL_SHA:-undefined}"
 
 COMMON_FLAGS="\
-    --incompatible_linkopts_to_linklibs \
+    --config=clang \
```
Comment (oschaaf) [inline maistra/run-ci.sh:19]: I wonder, can we do without this ` --config=clang`?

**Review by oschaaf — APPROVED** [review]
LGTM, awesome. Left a few questions, but nothing blocking.

File: bazel/envoy_internal.bzl (hunk 1 above)
Comment (jwendell) [reply bazel/envoy_internal.bzl:10]: We could, but I slightly prefer to keep those around, as they keep bothering us, thus motivating us to not forget about them.

File: bazel/foreign_cc/luajit.patch (hunk 2 above)
Comment (jwendell) [reply bazel/foreign_cc/luajit.patch:19]: yeah, for now I'm removing everything. We can work towards re-enabling them with the possible exception of annobin.

File: bazel/repositories.bzl (hunk 3 above)
Comment (jwendell) [reply bazel/repositories.bzl:826]: Kind of. The plan is to investigate which options are necessary. If we figure out we don't need any of them, we should get rid of this patch permanently.

File: maistra/bazelrc (hunk 4 above)
Comment (jwendell) [reply maistra/bazelrc:2]: We do. I had a hard time trying to workaround some warnings/errors (-Werror=xxx) in the proxy build. Adding back this flag (we had it in 2.1) is the way to move on with it.

File: maistra/run-ci.sh (hunk 5 above)
Comment (jwendell) [reply maistra/run-ci.sh:19]: Apparently there's a bazel build that does not identify correctly the compiler, which makes bazel/envoy think it's gcc and then enabling gcc-specific options. Then we get *lots* of warnings about unknown flags in clang. This basically declares the variable `BAZEL_COMPILER=clang`

Comment (oschaaf) [reply maistra/run-ci.sh:19]: Aaah I see, sounds like there's good a good reason for it to be there 👍

**Pull request merged**